### PR TITLE
Karazhan Chess Event - WIP

### PR DESCRIPTION
**Pull request by Gargarensis**
Just a work of translating [this](https://github.com/scriptdev2/scriptdev2-cata/commit/8f7573e0f5018961cbcf3586a2af5dc611189a00), for our core.
I'm putting it here even in its current terrible state so everyone can collaborate.

**Known issues and TODO list:**

- I think that some spells are missing scripts or are used improperly since they make the game crash.
(i.e. I had to use DoCastSpell instead of the Mangos' function DoCastSpellIfCan, since we do not have it)

- There is an **HORRIBLE** hack that **MUST** be adjusted regarding the addSC script made to avoid circular dependency with instance_karazhan.

**Tests performed:** It compiles with warnings, it still has to be adjusted and it is a 100% WIP.

### CODE REVIEW

**Comment by sucofog** [comment]
If you need help testing or with something tell me

**Comment by Yehonal** [comment]
@Gargarensis could you comment (here on github) lines that must be adapted/has issues? 

I'll check next days when i'll come back to AC

**Comment by Gargarensis** [comment]
Also, do not forget to set the scriptname:
```UPDATE creature_template SET ScriptName='npc_echo_of_medivh' WHERE entry=16816;
UPDATE creature_template SET ScriptName='npc_king_llane' WHERE entry=21684;
UPDATE creature_template SET ScriptName='npc_warchief_blackhand' WHERE entry=21752;
UPDATE creature_template SET ScriptName='npc_human_conjurer' WHERE entry=21683;
UPDATE creature_template SET ScriptName='npc_orc_warlock' WHERE entry=21750;
UPDATE creature_template SET ScriptName='npc_human_footman' WHERE entry=17211;
UPDATE creature_template SET ScriptName='npc_orc_grunt' WHERE entry=17469;
UPDATE creature_template SET ScriptName='npc_water_elemental' WHERE entry=21160;
UPDATE creature_template SET ScriptName='npc_summoned_daemon' WHERE entry=21726;
UPDATE creature_template SET ScriptName='npc_human_charger' WHERE entry=21664;
UPDATE creature_template SET ScriptName='npc_orc_wolf' WHERE entry=21748;
UPDATE creature_template SET ScriptName='npc_human_cleric' WHERE entry=21682;
UPDATE creature_template SET ScriptName='npc_orc_necrolyte' WHERE entry=21747;
UPDATE creature_template SET ScriptName='npc_chess_square' WHERE entry=17208;
UPDATE creature_template SET ScriptName='npc_chess_square' WHERE entry=17316;
```

**Comment by Yehonal** [comment]
Use our script to create pending sql and then push to this branch ;)

**Comment by Gargarensis** [comment]
I will do it as soon as possible, but here is the line of codes that make the first crash in the meantime:
```
            invoker->CastSpell(invoker, SPELL_DISABLE_SQUARE, true);
            invoker->CastSpell(invoker, SPELL_IS_SQUARE_USED, true);
```
line 355, 356 boss_chess_event.cpp

**Comment by Irome** [comment]
It would be a miracle if someone could fix this event.
Here are all the scripts I've found on the Chess Event, maybe they can be useful:

https://bitbucket.org/oregon/oregoncore/src/daad03585c3ed54209dcd01d475846fc85b0297a/src/scripts/EasternKingdoms/Karazhan/chess_event.cpp?at=default&fileviewer=file-view-default
https://bitbucket.org/lovelol/tbc/src/64e64b89ea60/Source/src/bindings/scripts/scripts/zone/karazhan/chess_event.cpp?fileviewer=file-view-default
https://pastebin.com/nCc0Rftj

**Comment by Yehonal** [comment]
@Gargarensis you can comment directly on those line in this PR, it's better than copy/paste it

**Comment by Irome** [comment]
Can anyone control this PR and help @Gargarensis to complete this event?

**Comment by sucofog** [comment]
Contact me by private @Gargarensis if you need any help about fixing the crashes

**Comment by Gargarensis** [comment]
@sucofog it would be great :) tell me how can i help you! 
I've put the code here so that everyone can contribute because i do not have a lot of time (and maybe not even the ability) to fix it by myself.

**Comment by talamortis** [comment]
i will work on this a little later in the week try get things rolling.

**Comment by BillytheOrc** [comment]
This would be kickass!

File: src/scripts/EasternKingdoms/Karazhan/boss_chess_event.cpp [hunk 1]
```diff
@@ -0,0 +1,1838 @@
...
+};
+
+/*######
+## npc_water_elemental
+######*/
+
+class npc_water_elemental : public CreatureScript
+{
+public:
+    npc_water_elemental() : CreatureScript("npc_water_elemental") { }
```
Comment (talamortis) [inline src/scripts/EasternKingdoms/Karazhan/boss_chess_event.cpp]: There is already a creature in the database called npc_water_elemental try change the script name to something like npc_kara_water_elemental

File: src/scripts/EasternKingdoms/Karazhan/instance_karazhan.cpp [hunk 2]
```diff
@@ -32,6 +32,21 @@ class instance_karazhan : public InstanceMapScript
...
+            m_uiChessResetTimer = 0;
+            m_uiAllianceStalkerCount = 0;
+            m_uiHordeStalkerCount = 0;
+        }
+
+        uint32 GetPlayerTeam() { return m_uiTeam; }
+        void GetChessPiecesByFaction(std::list<uint64>& lList, uint32 uiFaction) { lList = uiFaction == FACTION_ID_CHESS_ALLIANCE ? m_lChessPiecesAlliance : m_lChessPiecesHorde; };
+        Creature* GetCreatureByGUID(uint64 guid) { return instance->GetCreature(guid); };
+        
+        void OnPlayerEnter(Player* player)
```
Comment (talamortis) [inline src/scripts/EasternKingdoms/Karazhan/instance_karazhan.cpp]: void OnPlayerEnter(Player* player) override

**Comment by BillytheOrc** [comment]
@talamortis How's this going?

**Comment by talamortis** [comment]
still wip managed to get compile without errors or crashers jusrt need to get the event to work

**Comment by Irome** [comment]
is there any chance of being able to fix this PR?

**Comment by BarbzYHOOL** [comment]
You guys could try this, it's for TC I think and it was provided by @knindzagxg https://www78.zippyshare.com/v/o2MR1mGF/file.html

**Comment by BillytheOrc** [comment]
Any updates on this @BarbzYHOOL ?
That pull had a corrupt line 29 apparently.

**Comment by BarbzYHOOL** [comment]
No, @Knindzagxg has another version I think but it hasn't been coded by me or by him so someone would have to try to make it work manually first :(

**Comment by BarbzYHOOL** [comment]
Check this link guys, 3 scripts related to this issue https://mega.nz/fm/d3hFEaCT

**Comment by BillytheOrc** [comment]
All those links are dead @BarbzYHOOL

**Comment by BarbzYHOOL** [comment]
I think you need a mega account to see them but then it should work

**Comment by BarbzYHOOL** [comment]
Ok the link is https://mega.nz/#F!5vpTXQzL!2k7SuQCQMHl3MgCMSqogfg

**Comment by BillytheOrc** [comment]
is there any progress on this? maybe I can help?

**Comment by BarbzYHOOL** [comment]
@pak3935 no, but if you could help us test other Pull Requests, it would give more time to complete the very old pull requests

**Comment by FrancescoBorzi** [comment]
Looking for @Gargarensis or someone else to completed the work that has been previously started here.

**Comment by Viste** [comment]
i do it

**Comment by Viste** [comment]
Closed tomorrow open new PR based on this

**Pull request closed without merging**